### PR TITLE
Use STATIC instead of OBJECT for wasm-split-lib

### PR DESCRIPTION
**Pull request by tlively**
In an attempt to fix https://github.com/emscripten-core/emsdk/issues/886.

### CODE REVIEW

**Review by dschuff — APPROVED** [review]
ah, yeah I think this is what we want.
I think the other thing that would work might be to remove the `target_link_libraries` on line 10 and instead refer to the objects from the objlib directly in `binaryen_add_executable` (something like `instrumeter.cpp:wasm-split-lib`?). But this seems simpler.

**Comment by dschuff** [comment]
Oh, the other thing that https://cmake.org/cmake/help/latest/command/add_library.html says is that as of CMake 3.12 you actually *can* use object libraries in `target_link_libraries`. So that probably explains why it works everywhere except for emsdk.

**Comment by tlively** [comment]
aha, that does make sense. Anyway I don't have a problem with making a tiny static library here, so I'll just land this fix.

For my own edification, why would you choose to use the object files over the static library? IIUC, the library has an index that makes linking faster, but OTOH I suppose for a small enough library the time to create it might outweigh that benefit?

File: src/tools/wasm-split/CMakeLists.txt [hunk 1]
```diff
@@ -6,5 +6,5 @@ set(wasm_split_SOURCES
 )
 add_library(wasm-split-lib OBJECT ${wasm_split_SOURCES})
 
-binaryen_add_executable(wasm-split wasm-split.cpp)
+binaryen_add_executable(wasm-split wasm-split.cpp $<TARGET_OBJECTS:wasm-split-lib>)
 target_link_libraries(wasm-split wasm-split-lib)
```
Comment (sbc100) [inline src/tools/wasm-split/CMakeLists.txt:10]: I don't think it makes sense to add the object files from `wasm-split-lib` as wells the library itself.

Surely `target_link_libraries` entire purpose is to include the library so I don't see why this change would be needed?

File: src/tools/wasm-split/CMakeLists.txt [hunk 2]
```diff
@@ -6,5 +6,5 @@ set(wasm_split_SOURCES
 )
 add_library(wasm-split-lib OBJECT ${wasm_split_SOURCES})
```
Comment (sbc100) [inline src/tools/wasm-split/CMakeLists.txt:7]: Is there some reason we need `OBJECT` where?  I would have though `STATIC` makes more sense is the more common way to do things.

File: src/tools/wasm-split/CMakeLists.txt (hunk 1 above)
Comment (tlively) [reply src/tools/wasm-split/CMakeLists.txt:10]: Oh right, I forgot to remove the `target_link_libraries`.

File: src/tools/wasm-split/CMakeLists.txt (hunk 2 above)
Comment (tlively) [reply src/tools/wasm-split/CMakeLists.txt:7]: I tried STATIC, but it failed on the mingw builder. I was probably just holding it wrong, but this approach worked. Here's the failing log: https://github.com/WebAssembly/binaryen/runs/3539498459.

File: src/tools/wasm-split/CMakeLists.txt (hunk 1 above)
Comment (dschuff) [reply src/tools/wasm-split/CMakeLists.txt:10]: Yeah, my interpretation of the [doc](https://cmake.org/cmake/help/latest/command/add_library.html) is that if we use this TARGET_OBJECTS thing, it would be instead of target_link_libraries rather than in addition to it.

**Comment by dschuff** [comment]
> aha, that does make sense. Anyway I don't have a problem with making a tiny static library here, so I'll just land this fix.
> 
> For my own edification, why would you choose to use the object files over the static library? IIUC, the library has an index that makes linking faster, but OTOH I suppose for a small enough library the time to create it might outweigh that benefit?

Aside from the index, linking an archive (rather than including an object file directly on the link line) means that an object file is only added to the link if it satisfies some undefined reference to one of its symbols (whereas when the object goes on the link line directly, it gets forced in regardless). I think if we're going to use a library, it might as well be a static library unless there's some good reason not to.

File: src/tools/wasm-split/CMakeLists.txt (hunk 2 above)
Comment (dschuff) [reply src/tools/wasm-split/CMakeLists.txt:7]: Ah, I'll bet that's a link order issue. When using GNU ld, each archive is checked only once, in the order they appear on the linker command line. If an object file is needed to satisfy an undefined reference, it gets added to the link. But if another object is added later and has an undefined reference to an object in the earlier archive (that wasn't previously added), then you can get this kind of error. lld and link.exe will basically iterate over all the libraries until convergence.

So to fix this, we could fix the link order of the libraries (or fix any legit circular dependencies), or we can just give ld the same behavior as the others by adding `--start-group` and `--end-group` around the relevant sets of libraries. I wonder if there's an easy way to make CMake do that...

**Comment by tlively** [comment]
Ok, trying a fix in a new PR: https://github.com/WebAssembly/binaryen/pull/4132.
